gdx/src/com/badlogic/gdx/graphics/g3d/shaders/DefaultShader.java

```diff
@@ -646,75 +646,75 @@ private static final long combineAttributeMasks (final Renderable renderable) {
-			prefix += "#define " + TextureAttribute.EmissiveAlias + "Flag\n";
-			prefix += "#define " + TextureAttribute.EmissiveAlias + "Coord texCoord0\n"; // FIXME implement UV mapping
+			prefix.append("#define " + TextureAttribute.EmissiveAlias + "Flag\n");
+			prefix.append("#define " + TextureAttribute.EmissiveAlias + "Coord texCoord0\n"); // FIXME implement UV mapping
 		}
 		if ((attributesMask & TextureAttribute.Reflection) == TextureAttribute.Reflection) {
-			prefix += "#define " + TextureAttribute.ReflectionAlias + "Flag\n";
-			prefix += "#define " + TextureAttribute.ReflectionAlias + "Coord texCoord0\n"; // FIXME implement UV mapping
+			prefix.append("#define " + TextureAttribute.ReflectionAlias + "Flag\n");
+			prefix.append("#define " + TextureAttribute.ReflectionAlias + "Coord texCoord0\n"); // FIXME implement UV mapping
 		}
 		if ((attributesMask & TextureAttribute.Ambient) == TextureAttribute.Ambient) {
-			prefix += "#define " + TextureAttribute.AmbientAlias + "Flag\n";
-			prefix += "#define " + TextureAttribute.AmbientAlias + "Coord texCoord0\n"; // FIXME implement UV mapping
+			prefix.append("#define " + TextureAttribute.AmbientAlias + "Flag\n");
+			prefix.append("#define " + TextureAttribute.AmbientAlias + "Coord texCoord0\n"); // FIXME implement UV mapping
 		}
 		if ((attributesMask & ColorAttribute.Diffuse) == ColorAttribute.Diffuse)
-			prefix += "#define " + ColorAttribute.DiffuseAlias + "Flag\n";
+			prefix.append("#define " + ColorAttribute.DiffuseAlias + "Flag\n");
 		if ((attributesMask & ColorAttribute.Specular) == ColorAttribute.Specular)
-			prefix += "#define " + ColorAttribute.SpecularAlias + "Flag\n";
+			prefix.append("#define " + ColorAttribute.SpecularAlias + "Flag\n");
 		if ((attributesMask & ColorAttribute.Emissive) == ColorAttribute.Emissive)
-			prefix += "#define " + ColorAttribute.EmissiveAlias + "Flag\n";
+			prefix.append("#define " + ColorAttribute.EmissiveAlias + "Flag\n");
 		if ((attributesMask & ColorAttribute.Reflection) == ColorAttribute.Reflection)
-			prefix += "#define " + ColorAttribute.ReflectionAlias + "Flag\n";
+			prefix.append("#define " + ColorAttribute.ReflectionAlias + "Flag\n");
 		if ((attributesMask & FloatAttribute.Shininess) == FloatAttribute.Shininess)
-			prefix += "#define " + FloatAttribute.ShininessAlias + "Flag\n";
+			prefix.append("#define " + FloatAttribute.ShininessAlias + "Flag\n");
 		if ((attributesMask & FloatAttribute.AlphaTest) == FloatAttribute.AlphaTest)
-			prefix += "#define " + FloatAttribute.AlphaTestAlias + "Flag\n";
-		if (renderable.bones != null && config.numBones > 0) prefix += "#define numBones " + config.numBones + "\n";
-		return prefix;
+			prefix.append("#define " + FloatAttribute.AlphaTestAlias + "Flag\n");
+		if (renderable.bones != null && config.numBones > 0) prefix.append("#define numBones ").append(config.numBones).append("\n");
+		return prefix.toString();

```

Comment:
This whole thing: This does not make the code less readable IMO. And since this code runs for each new `DefaultShader` creation, I'd probably be more comfortable knowing that it does not generate 200 strings. Most of the lines are conditional and will typically not appear, but still.